---
title: "Graded Problem Set: Investment and Portfolio Management"
subtitle: "Imperial College London - Business School"
author: 
  - name: "Rodolphe Lajugie"
date: "2025-10-31"
format: pdf
fontsize: 12pt
geometry: margin=1in
titlepage: true
titlepage-geometry: "top=2cm, bottom=2cm, left=2cm, right=2cm"
---
\newpage
\tableofcontents
\listoffigures
\newpage
\listoftables
\newpage

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import itertools

In [2]:
df = pd.read_excel('../data/data_coursework1_Q1.xlsx')
df = df[['year_', 'month_', 'date_', '1-month Tbill', 'SP500', 'IBM']]

# Calculate returns
df['SP500_ret'] = df['SP500'].pct_change()
df['IBM_ret'] = df['IBM'].pct_change()

# Adjust T-bill to monthly rate and scale (if needed)
df['rft'] = df['1-month Tbill'] / 100

df = df.dropna(subset=['SP500_ret', 'IBM_ret', 'rft'])

# Excess returns
df['excess_stock'] = df['IBM_ret'] - df['rft']
df['excess_market'] = df['SP500_ret'] - df['rft']
df['date'] = pd.to_datetime(df['year_'].astype(
    str) + '-' + df['month_'].astype(str), format='%Y-%m')
df.drop(columns=['year_', 'month_', 'date_'], inplace=True)
df

1-month Tbill   SP500    IBM  SP500_ret   IBM_ret     rft  excess_stock  \
25            0.20   70.22   2.66   0.016650 -0.011152  0.0020     -0.013152   
26            0.20   70.29   2.64   0.000997 -0.007519  0.0020     -0.009519   
27            0.22   68.05   2.25  -0.031868 -0.147727  0.0022     -0.149927   
28            0.24   62.99   1.95  -0.074357 -0.133333  0.0024     -0.135733   
29            0.20   55.63   1.68  -0.116844 -0.138462  0.0020     -0.140462   
..             ...     ...    ...        ...       ...     ...           ...   
367           0.66  330.70  19.89  -0.081389 -0.076173  0.0066     -0.082773   
368           0.60  315.40  20.77  -0.046265  0.044243  0.0060      0.038243   
369           0.68  307.10  20.60  -0.026316 -0.008185  0.0068     -0.014985   
370           0.57  315.20  22.43   0.026376  0.088835  0.0057      0.083135   
371           0.60  328.70  22.31   0.042830 -0.005350  0.0060     -0.011350   

     excess_market       date  
25        0.014650 1962-02-01  
26       -0.001003 1962-03-01  
27       -0.034068 1962-04-01  
28       -0.076757 1962-05-01  
29       -0.118844 1962-06-01  
..             ...        ...  
367      -0.087989 1990-08-01  
368      -0.052265 1990-09-01  
369      -0.033116 1990-10-01  
370       0.020676 1990-11-01  
371       0.036830 1990-12-01  

[347 rows x 9 columns]

# Model 1: Simple CAPM Regression

## Part 1: linear regression
### Manual computation

We know that $\hat{\beta} = (X'X)^{-1}X'y$ <br>
Where X is the design matrix and y the response variable.


In [3]:
y = df['excess_stock'].values
X = np.column_stack((np.ones(len(df)), df['excess_market'].values))

In [4]:
beta_hat = np.linalg.inv(X.T @ X) @ (X.T @ y)
alpha, beta = beta_hat
print('alpha:', alpha, 'beta:', beta)

alpha: 0.0027020468479849896 beta: 0.8856018941150132


### With builtin functions

In [5]:
X1 = sm.add_constant(df['excess_market'])
model1 = sm.OLS(df['excess_stock'], X1).fit()
print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:           excess_stock   R-squared:                       0.286
Model:                            OLS   Adj. R-squared:                  0.284
Method:                 Least Squares   F-statistic:                     138.4
Date:                Tue, 18 Nov 2025   Prob (F-statistic):           4.24e-27
Time:                        17:14:25   Log-Likelihood:                 543.44
No. Observations:                 347   AIC:                            -1083.
Df Residuals:                     345   BIC:                            -1075.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const             0.0027      0.003      0.993

Using the OLS function from the statsmodels library or by computing it manually, we obtain the following results for the CAPM regression of IBM's excess returns against the market's excess returns:
- Intercept ($\alpha$): 0.0027
- Slope ($\beta$): 0.8856

$\alpha$ is close to zero which means IBM doesn't outperform or underperform the market on average. The $\beta$ of 0.89 indicates that IBM is less volatile than the market.

## Part 2: 2nd linear regression

We want to find $\alpha$, $\beta_1$, $\beta_2$ and $\beta_3$ such that: <br>
$$r_{it}-r_{ft} = \alpha + \beta_1 [D_t(r_{mt}-r_{ft})] + \beta_2[(1-D_t)(r_{mt}-r_{ft})] + \beta_3 (r_{mt}-r_{ft})^2 + u_t$$

Where $D_t = 1 $ if the market excess return is positive and 0 otherwise.

### Manual computation

In [6]:
Dt = (df['excess_market'] > 0).astype(int).values
D_excess = Dt * df['excess_market'].values
nonD_excess = (1 - Dt) * df['excess_market'].values
excess_market_sq = df['excess_market'].values ** 2

Again using the formula $\hat{\beta} = (X'X)^{-1}X'y$ we obtain:

In [7]:
X2 = np.column_stack(
    (np.ones(len(df)), D_excess, nonD_excess, excess_market_sq))
beta_hat2 = np.linalg.inv(X2.T @ X2) @ (X2.T @ y)
print('Model 2 coefficients:\nalpha: ', beta_hat2[0], ' beta1: ', beta_hat2[1],
      '\nbeta2: ', beta_hat2[2], ' beta3: ', beta_hat2[3])

Model 2 coefficients:
alpha:  -0.010031847996995746  beta1:  1.6230560811213248 
beta2:  0.11105859440112376  beta3:  -6.095601716965916


### With builtin functions

In [8]:
df['Dt'] = (df['excess_market'] > 0).astype(int)
df['D_excess'] = df['Dt'] * df['excess_market']
df['nonD_excess'] = (1 - df['Dt']) * df['excess_market']
df['excess_market_sq'] = df['excess_market']**2

df.head()

1-month Tbill  SP500   IBM  SP500_ret   IBM_ret     rft  excess_stock  \
25           0.20  70.22  2.66   0.016650 -0.011152  0.0020     -0.013152   
26           0.20  70.29  2.64   0.000997 -0.007519  0.0020     -0.009519   
27           0.22  68.05  2.25  -0.031868 -0.147727  0.0022     -0.149927   
28           0.24  62.99  1.95  -0.074357 -0.133333  0.0024     -0.135733   
29           0.20  55.63  1.68  -0.116844 -0.138462  0.0020     -0.140462   

    excess_market       date  Dt  D_excess  nonD_excess  excess_market_sq  
25       0.014650 1962-02-01   1   0.01465     0.000000          0.000215  
26      -0.001003 1962-03-01   0  -0.00000    -0.001003          0.000001  
27      -0.034068 1962-04-01   0  -0.00000    -0.034068          0.001161  
28      -0.076757 1962-05-01   0  -0.00000    -0.076757          0.005892  
29      -0.118844 1962-06-01   0  -0.00000    -0.118844          0.014124

In [9]:
X3 = sm.add_constant(df[['D_excess', 'nonD_excess', 'excess_market_sq']])
model2 = sm.OLS(df['excess_stock'], X3).fit()
print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:           excess_stock   R-squared:                       0.299
Model:                            OLS   Adj. R-squared:                  0.293
Method:                 Least Squares   F-statistic:                     48.87
Date:                Tue, 18 Nov 2025   Prob (F-statistic):           2.53e-26
Time:                        17:14:25   Log-Likelihood:                 546.66
No. Observations:                 347   AIC:                            -1085.
Df Residuals:                     343   BIC:                            -1070.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -0.0100      0.006  

The second model allows different market exposures depending on whether excess market returns are positive or negative. The high value for beta1 relative to beta2 suggests IBM’s response to positive market movements is much larger than to negative movements. The negative beta3 indicates a possible concavity in the relationship, showing diminishing marginal response for larger market returns. <br>
$R^2$ has improved slightly from the first model, indicating a better fit.

## F Test

Here, we want to test the null hypothesis $H_0: \beta_1 = \beta_2$

### Manual computation

$\hat{y} = X\hat{\beta}$



(((($RSS = \sum_{i=1}^{n}(y_i - \hat{y_i})^2$))))

In [10]:
R= np.array([[0, 1, -1,0]])
r = 0

var1 = np.linalg.inv(X2.T @ X2)
var_R = np.linalg.inv(R @ var1 @ R.T)

e = df['excess_stock'].values - X2 @ beta_hat2
T = X2.shape[0]
e_eprime = (e.T @ e)/(T-4)

F_stats = ((R @ beta_hat2 - r)*var_R * (R @ beta_hat2 - r).T) / 1 / e_eprime
print('F statistic:', F_stats)

F statistic: [[5.76675586]]


In [11]:
from scipy import stats

stats.f.sf(F_stats, 1, T - 4)

array([[0.01686377]])

### With builtin functions

In [12]:
f_test_result = model2.f_test('D_excess = nonD_excess')
print(f_test_result)

<F test: F=5.766755863532396, p=0.01686377114880381, df_denom=343, df_num=1>


## T-test 

In [13]:
y_hat1 = X @ beta_hat
residuals1 = y - y_hat1
sigma2 = np.sum(residuals1 ** 2) / (len(y) - X.shape[1])
cov_beta_hat = sigma2 * np.linalg.inv(X.T @ X)
se_alpha = np.sqrt(cov_beta_hat[0, 0])
t_stat_alpha = alpha / se_alpha
print('t statistic for alpha:', t_stat_alpha)

t statistic for alpha: 0.9930805477699152


In [14]:
t_test_alpha = model1.t_test('const = 0')
print(t_test_alpha)

                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0             0.0027      0.003      0.993      0.321      -0.003       0.008


In [15]:
t_test_alpha2 = model2.t_test('const = 0')
print(t_test_alpha2)

                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0            -0.0100      0.006     -1.757      0.080      -0.021       0.001


# Question 2

In [16]:
df = pd.read_excel('../data/data_coursework1_Q2.xlsx', sheet_name="final_m")

Let's use  EMU stock total excess return as our Rt as defined in the question.
We can then compute Y as follows:
$$Y_t = \begin{cases} 1 & \text{if } R_{t} > 0 \\ 0 & \text{otherwise} \end{cases}$$

In [60]:
R = df['ESR']
df['Y'] = (R > 0).astype(int)

predictor_candidates = ['USR', 'EIF', 'ERECB',
                        'EFX', 'EDY', 'ESP', 'UIF', 'URR', 'UDY', 'UOIL']

df['Y_next'] = df['Y'].shift(-1).dropna()
# data = df.dropna().copy() #because shift

The regression model is then defined as: <br>
$$Y_{t+1} = \beta'X_t + \epsilon_{t}$$

In [63]:
X = df[predictor_candidates][:-1].values
y = df['Y_next'][:-1].values


XtX = X.T @ X
XtX_inv = np.linalg.inv(XtX)
XtX_inv @ X.T @ y

array([  1.12120012,   7.61779329,  -0.53160374,   0.9661701 ,
         0.95065933,  -0.04133255, -13.32873761,   0.4950263 ,
        -0.07638746,  -0.9665826 ])

In [52]:
beta_hat

array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan])

In [33]:
df['Y'] = (df['ESR'] > 0).astype(int)
df['Y_next'] = df['Y'].shift(-1)

predictors = ['USR', 'EIF', 'ERECB', 'EFX',
              'EDY', 'ESP', 'UIF', 'URR', 'UDY', 'UOIL']

# lagged
for col in predictors:
    df[col + '_lag'] = df[col].shift(1)

# cleaned dataset
data = df.dropna().copy()

X = data[[p + '_lag' for p in predictors]]
y = data['Y_next']

# OLS estimation
model_full = sm.OLS(y, X).fit()
print(model_full.summary())

                                 OLS Regression Results                                
Dep. Variable:                 Y_next   R-squared (uncentered):                   0.622
Model:                            OLS   Adj. R-squared (uncentered):              0.608
Method:                 Least Squares   F-statistic:                              45.90
Date:                Tue, 18 Nov 2025   Prob (F-statistic):                    3.01e-53
Time:                        17:35:52   Log-Likelihood:                         -191.12
No. Observations:                 289   AIC:                                      402.2
Df Residuals:                     279   BIC:                                      438.9
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

About the Z score computation, we define a function compute_Z such that:

$$ Z_t(\alpha) =  \begin{cases} 1 & \text{if } \hat{P}(R_t > 0 \text{ | } X_{t-1}) > \alpha \text{ and } R_t > 0 \\ 1 & \text{if } \hat{P}(R_t > 0 \text{ | } X_{t-1}) \leq \alpha \text{ and } R_t \leq 0 \\ 0 & \text{Otherwise} \end{cases} $$

For a fixed $\alpha \in \text{ ]0;1[ }$

$$Z(\alpha) = \frac{\sum_{t=2}^{n} Z_t(\alpha)}{n-1}$$

Where TPR is the true positive rate and FPR the false positive rate.


In [19]:
def compute_Z(y_true, p_hat, alpha):
    Z = np.where((p_hat > alpha) & (y_true == 1), 1,
                 np.where((p_hat <= alpha) & (y_true == 0), 1, 0))
    return Z.mean()

In [20]:
best_Z = -1 #initialize so that any Z_val is better but can't be worse
best_predictors = None
results = []

for k in range(1, 11): #all combos possible of size 1 to 10
    for subset in itertools.combinations(predictor_candidates, k):

        X = data[list(subset)]
        y = data['Y_next']

        model = sm.OLS(y, X).fit()
        p_hat = model.predict(X)  

        Z_val = compute_Z(y, p_hat, alpha=0.5) #first set (question specify it)
        results.append((subset, Z_val))

        if Z_val > best_Z:
            best_Z = Z_val
            best_predictors = subset

In [21]:
#| label : tbl-result-Z
#| tbl-cap : "Top 10 combinations of predictors by Z score"
df_results = pd.DataFrame([
    {
        "Subset": ", ".join(subset),
        "Subset_size": len(subset),
        "Z_score": z
    }
    for subset, z in results
])

# Tri du plus grand au plus petit Z
df_results = df_results.sort_values(
    "Z_score", ascending=False).reset_index(drop=True)

# Affichage
df_results.head(10)

Subset  Subset_size   Z_score
0        USR, ERECB, EFX, EDY, ESP, UIF, URR, UDY            8  0.667820
1            USR, EIF, ERECB, EDY, UIF, URR, UOIL            7  0.664360
2       USR, EIF, ERECB, EDY, ESP, UIF, URR, UOIL            8  0.664360
3                  USR, ERECB, EFX, EDY, UIF, URR            6  0.664360
4        USR, EIF, ERECB, EFX, EDY, UIF, URR, UDY            8  0.664360
5  USR, EIF, ERECB, EFX, EDY, ESP, UIF, URR, UOIL            9  0.660900
6         USR, EIF, EFX, EDY, ESP, UIF, URR, UOIL            8  0.660900
7             USR, EIF, ERECB, EFX, EDY, UIF, URR            7  0.660900
8            USR, ERECB, EFX, EDY, UIF, URR, UOIL            7  0.660900
9       USR, ERECB, EDY, ESP, UIF, URR, UDY, UOIL            8  0.657439

In [22]:
print("Best predictor combination (alpha = 0.5):\n", best_predictors)
print("Achieved Z(alpha):", best_Z)

Best predictor combination (alpha = 0.5):
 ('USR', 'ERECB', 'EFX', 'EDY', 'ESP', 'UIF', 'URR', 'UDY')
Achieved Z(alpha): 0.6678200692041523


In [23]:

X_best = data[list(best_predictors)]
y = data['Y_next']

model = sm.OLS(y, X_best).fit()
p_hat = model.predict(X_best)

alphas = np.linspace(0.00001, 0.99999, 99999) #increase for better precision but takes time
Z_list = [compute_Z(y, p_hat, a) for a in alphas]

alpha_best = alphas[np.argmax(Z_list)]
Z_best_alpha = max(Z_list)

print("Optimal alpha:", alpha_best)
print("Z(alpha*) =", Z_best_alpha)

Optimal alpha: 0.50102
Z(alpha*) = 0.6747404844290658
